File: leetcode/Python/0050-m-powx-n.py
"""
Recursion
Time: O(logn)
Space: O(logn)
"""
class Solution_1:
    def myPow(self, x: float, n: int) -> float:
        if n == 0:
            return 1
        if n < 0:
            n = -n
            x = 1 / x
        return self.myPow(x * x, n // 2) if n % 2 == 0 else x * self.myPow(x * x, n // 2)

"""
Binary a^(1010)2 = a^(1000)2 * a^(10)2
Time: O(logn)
Space: O(1)
"""
class Solution_2:
    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            x = 1 / x
            n = -n
        res = 1
        tmp = x # x^1, x^2, x^4, x^8...
        while n != 0:
            if n % 2 == 1:
                res *= tmp
            tmp *= tmp
            n //= 2
        return res

File: leetcode/C++/0050-m-powx-n.cpp
// Time: O(logn)
// Space: O(logn)
class Solution {
public:
    double helper(double x, long long n) {
        if (n == 0)
            return 1.0;
        return n % 2 == 0 ? helper(x * x, n / 2) : x * helper(x * x, n / 2);
    }
    
    double myPow(double x, int n) {
        long long N = n;
        if (N < 0) {
            x = 1 / x;
            N = -N;
        }
        return helper(x, N);
    }
};

// Iterative
// Time: O(logn)
// Space: O(1)
class Solution {
public:
    double myPow(double x, int n) {
        long long N = n;
        if (N < 0) {
            x = 1 / x;
            N = -N;
        }
        double res = 1;
        double cur_product = x;
        for (long long i = N; i; i /= 2) {
            if ((i % 2) == 1) {
                res = res * cur_product;
            }
            cur_product *= cur_product;
        }
        return res;
    }
};